src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java

```diff
@@ -79,19 +83,97 @@
+     * actions is added, which sets the custom type (and all it's fields to the old resource).</li>
+     * <li>If the new resource's custom type is not set, but the old resource's custom type is set. A
+     * "setCustomType" update action is added, which removes the type set on the old resource.</li>
+     * <li>If both the resources custom types are the same and the custom fields are both set. The custom
+     * field values of both resources are then calculated. (see
+     * {@link CustomUpdateActionUtils#buildSetCustomFieldsUpdateActions(Map, Map, Custom, Class, Integer, Function,
+     * Function, Function, BaseSyncOptions)})</li>
+     * <li>If the keys of both custom types are different, then a "setCustomType" update action is added, where the
+     * old resource's custom type is set to be as the new one's.</li>
+     * <li>If both resources custom type keys are identical but the custom fields of the new resource's custom type is
+     * not set.</li>
+     * </ol>
+     *
+     * <p>An update action will <b>not</b> be added to the result list in the following cases:-
+     * <ol>
+     * <li>If both the resources' custom types are not set.</li>
+     * <li>If both the resources' custom type keys are not set.</li>
+     * <li>Custom fields are both empty.</li>
+     * <li>Custom field JSON values have different ordering.</li>
+     * <li>Custom field values are identical.</li>
+     * </ol>
+     *
+     * @param <T> the type of the old {@link Resource} which has the custom fields.
+     * @param <S> the type of the new resource {@link CustomDraft}.
+     * @param <U> the type of the resource in which the update actions will be applied on.
+     *
+     * @param oldResource the resource which should be updated.
+     * @param newResource the resource draft where we get the new custom fields.
+     * @param containerResourceClass the class of the container resource which will be updated.
+     * @param variantId optional field representing the variant id in case the oldResource is an asset.
+     * @param resourceIdGetter a function used to get the id of the resource being updated.
+     * @param resourceTypeIdGetter a function used to get the Type id of the resource being updated.
+     * @param updateIdGetter a function used to get the id/key needed for updating the resource that has the custom
+     *                       fields.
+     * @param syncOptions responsible for supplying the sync options to the sync utility method.
+     * @return a list that contains all the update actions needed, otherwise an
+     *      empty list if no update actions are needed.
+     */
+    @Nonnull
+    public static <T extends Custom, S extends CustomDraft, U extends Resource<U>> List<UpdateAction<U>>
         buildCustomUpdateActions(
         @Nonnull final T oldResource,
         @Nonnull final S newResource,
+        @Nullable final Class<U> containerResourceClass,
```

Comment:
That's actually where the enum instances i complained above (`ConcreteBuilder`) could be useful: instead of operating with classes and throwing/handling unchecked exceptions, prevent unsupported values passing to to the builder!

How library consumer (or even library developer/collaborator/contributor) should know which types are supported when calling this method, besides runtime exception handling? 
`buildCustomUpdateActions -> buildTypedSetCustomTypeUpdateAction -> GenericCustomActionBuilderFactory.of -> GenericCustomActionBuilderFactory.getBuilder -> ConcreteBuilder` this is the chain i had to go to answer this question.

Reply:
Now the [`GenericCustomActionBuilderFactory`](https://github.com/commercetools/commercetools-sync-java/pull/253/files#diff-5bf1199b0b3b82e94b89c6a5a0f12c16) is removed at all. And now the concrete implementations are [passed to the utils](https://github.com/commercetools/commercetools-sync-java/pull/253/files#diff-cea7e36a8d3c43856febd3c4d36f8145R59).